Frameworks/CoreGraphics/CGColor.mm

```diff
@@ -17,37 +17,149 @@
+            typeid(__CGColor).name(), // Class Name
+            nullptr, // Init
+            nullptr, // Copy (ill-defined for CF)
+            &_CFFinalize, // Finalize
+            &__CFEquals, // Equals
+            nullptr, // Hash
+            nullptr, // Copy Description
+            nullptr, // Copy Debug Description
+        };
+        static CFTypeID _id = _CFRuntimeRegisterClass(&_cls);
+        return _id;
+    }
+
+    __CGColor(CGColorSpaceRef colorspace, CGFloat r, CGFloat g, CGFloat b, CGFloat alpha)
+        : _colorSpace(colorspace), _pattern(nullptr), _components{ r, g, b, alpha } {
+    }
+
+    __CGColor(CGColorSpaceRef colorSpace, CGPatternRef pattern) : _colorSpace(colorSpace), _pattern(pattern) {
+        // TODO #2581: CGColorCreateWithPattern should support components
+    }
+
+    // CreateCopyWithAlpha
+    __CGColor(const __CGColor& color, float alpha) : _colorSpace(color._colorSpace), _pattern(color._pattern) {
+        if (color.IsPatternColor()) {
+            ClearColorComponents();
+            _components.a = alpha;
+        } else {
+            SetColorComponents(color._components.r, color._components.g, color._components.b, alpha);
+        }
+    }
+
+    bool operator==(const __CGColor& other) const {
+        if ((CGColorSpaceGetModel(_colorSpace) != CGColorSpaceGetModel(other._colorSpace)) || (_pattern != other._pattern) ||
+            !(_components == other._components)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    inline CGFloat Alpha() const {
+        return _components.a;
+    }
+
+    inline CGColorSpaceRef ColorSpace() const {
+        return _colorSpace;
+    }
+
+    inline const __CGColorQuad& ColorComponents() const {
+        return _components;
+    }
+
+    inline CGPatternRef Pattern() const {
+        return _pattern;
+    }
+
+private:
+    __CGColorQuad _components;
+    woc::StrongCF<CGColorSpaceRef> _colorSpace;
+    woc::StrongCF<CGPatternRef> _pattern;
+
+    static Boolean __CFEquals(CFTypeRef color1, CFTypeRef color2) {
+        return *((__CGColor*)color1) == *((__CGColor*)color2);
+    }
+
+    inline void ClearColorComponents() {
+        _components.r = 0.0f;
+        _components.g = 0.0f;
+        _components.b = 0.0f;
+        _components.a = 0.0f;
+    }
+
+    inline void SetColorComponents(CGFloat r, CGFloat g, CGFloat b, CGFloat alpha) {
+        _components.r = r;
+        _components.g = g;
+        _components.b = b;
+        _components.a = alpha;
+    }
+
+    inline bool IsPatternColor() const {
+        return _pattern != nullptr;
+    }
+};
+
+/**
+ @Status Caveat
+ @Notes components are not supported
+*/
+CGColorRef CGColorCreateWithPattern(CGColorSpaceRef colorSpace, CGPatternRef pattern, const CGFloat* components) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!pattern);
+    return __CGColor::CreateInstance(kCFAllocatorDefault, colorSpace, pattern);
+}
+
+static inline CGColorRef __CreateBlackColor() {
```

Comment:
nit: could just inline these in CGColorGetConstantColor and share a static colorspace